File: Breadth-First Search/0490.The Maze/Solution.py
class Solution:
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
        m, n = len(maze), len(maze[0])
        vis = [[False] * n for _ in range(m)]
        q = deque()
        q.append(start)
        vis[start[0]][start[1]] = True
        dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        while q:
            cur = q.popleft()
            for d in dirs:
                nx, ny = cur[0] + d[0], cur[1] + d[1]
                while 0 <= nx < m and 0 <= ny < n and maze[nx][ny] == 0:
                    nx += d[0]
                    ny += d[1]
                nx -= d[0]
                ny -= d[1]
                if vis[nx][ny]:
                    continue
                vis[nx][ny] = True
                q.append([nx, ny])
        return vis[destination[0]][destination[1]]

File: Breadth-First Search/0490.The Maze/Solution.java
class Solution {
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        int m = maze.length, n = maze[0].length;
        boolean[][] vis = new boolean[m][n];
        Deque<int[]> q = new ArrayDeque<>();
        q.add(start);
        vis[start[0]][start[1]] = true;
        int[][] dirs = new int[][]{{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        while (!q.isEmpty()) {
            int[] cur = q.pollFirst();
            for (int[] d : dirs) {
                int nx = cur[0] + d[0], ny = cur[1] + d[1];
                while (nx >= 0 && nx < m && ny >= 0 && ny < n && maze[nx][ny] == 0) {
                    nx += d[0];
                    ny += d[1];
                }
                nx -= d[0];
                ny -= d[1];
                if (vis[nx][ny]) {
                    continue;
                }
                vis[nx][ny] = true;
                q.addLast(new int[]{nx, ny});
            }
        }
        return vis[destination[0]][destination[1]];
    }
}